### ISSUE TITLE
Memory corruption bugs in Triton kernels

### ISSUE THREAD

**Issue by eusuntbogdan** [thunlp/TritonBench#12]
## What broke

We ran the standard tests in `data/TritonBench_G_v1` using multiple memory-safety sanitizers and identified several memory corruption bugs in the Triton kernels.

After manual analysis, we confirmed **24 true-positive reports** affecting the following files:

* `apply_penalty.py`
* `attention_fwd_triton1.py`
* `chunk_bwd_dqkg.py`
* `chunk_delta_fwd.py`
* `chunk_gla_fwd.py`
* `chunk_gla_simple.py`
* `decay_cumsum.py`
* `fused_rotary_embedding.py`
* `matmul_triton2.py`
* `nested_loops_processing.py`
* `parallel_retention_attention.py`
* `chunk_gated_attention.py`
* `chunk_linear_attn.py`
* `chunk_retention_ops.py`
* `fused_recurrent_delta.py`
* `fused_recurrent_hgrn.py`
* `fused_recurrent_retention.py`
* `layer_norm_triton.py`
* `parallel_attention.py`
* `rotary_emb_nopad.py`
* `token_attn_mistral.py`
* `token_attn_reduceV.py`
* `quantize_kv_copy.py`
* `softmax_reducev.py`

## Reproduction

Each file contains a standard test that can be run directly from the TritonBench repository root. For example:

```bash
python data/TritonBench_G_v1/apply_penalty.py
```

NVIDIA Compute Sanitizer reports invalid memory accesses in the following files:

* `chunk_gla_fwd.py`
* `chunk_linear_attn.py`
* `chunk_retention_ops.py`
* `parallel_attention.py`
* `rotary_emb_nopad.py`
* `token_attn_mistral.py`
* `token_attn_reduceV.py`
* `softmax_reducev.py`

For example:

```bash
compute-sanitizer python data/TritonBench_G_v1/token_attn_mistral.py
```

Triton-Sanitizer reports invalid accesses in the following files when running the same standard tests:

* `apply_penalty.py`
* `attention_fwd_triton1.py`
* `chunk_bwd_dqkg.py`
* `chunk_delta_fwd.py`
* `chunk_gla_fwd.py`
* `chunk_gla_simple.py`
* `decay_cumsum.py`
* `fused_rotary_embedding.py`
* `matmul_triton2.py`
* `nested_loops_processing.py`
* `parallel_retention_attention.py`
* `quantize_kv_copy.py`

The remaining reports were identified by a sanitizer we are currently developing which we will open source soon using the same standard tests.

## Environment

* Python: 3.12.3
* Triton: 3.6.0
* PyTorch: 2.11
* CUDA: 13.0
* GPU: NVIDIA GB10

### `token_attn_mistral.py` and `token_attn_reduceV.py`

The root cause of the memory corruption in these two kernels is a division by zero while computing an index used for memory accesses.

The wrapper computes the grouping factor as:

```python
kv_group_num = prob.shape[0] // v.shape[1]
```

For the failing test case:

```text
prob.shape[0] = 4
v.shape[1] = 128
kv_group_num = 4 // 128 = 0
```

The kernel subsequently computes:

```python
cur_kv_head = cur_head // kv_group_num
```

On the Blackwell GPU used for testing, we observed division by zero in 32-bit integer arithmetic produce `2^31 - 1`. This value is then multiplied by the `V` head stride and used as part of the load address, resulting in a very large out-of-bounds access.

In short, the failure sequence is:

```text
v.shape[1] is the sequence length rather than a KV-head count
        ↓
kv_group_num == 0
        ↓
division by zero during index calculation
        ↓
invalid index (observed as 2^31 - 1)
        ↓
out-of-bounds load from V
```

The grouping factor should be derived from the actual number of KV heads, or the wrapper should reject launches for which `kv_group_num == 0`.

### Non-contiguous tensor bugs

The following kernels pass broadcasted, non-contiguous gradient tensors to backward Triton kernels:

| File | Densely indexed gradient argument(s) |
| --- | --- |
| `chunk_linear_attn.py` | `do` |
| `chunk_retention_ops.py` | `do` |
| `fused_recurrent_delta.py` | `do`, and `dht` when the final-state output participates in the loss |
| `fused_recurrent_hgrn.py` | `do` |
| `fused_recurrent_retention.py` | `do` |
| `layer_norm_triton.py` | `dy` |
| `parallel_attention.py` | `do` and `dz` |

The standard tests form scalar losses using operations such as `loss = o.sum()` and then call `loss.backward()`. PyTorch can represent the resulting output gradient as an expanded zero-stride view: it has the logical shape of the output, but its backing storage contains only one physical element.

The backward kernels do not use the actual strides of these gradient tensors. Instead, they index the gradients as dense tensors, commonly by reusing the strides of another input such as `q`, `v`, or `x`. Consequently, the first nonzero logical offset already points beyond the single element backing the expanded gradient.

In short, the failure sequence is:

```text
scalar loss created with sum()
        ↓
autograd supplies a broadcasted zero-stride gradient
        ↓
logical tensor extent is larger than its backing storage
        ↓
Triton kernel assumes a dense layout and ignores the gradient strides
        ↓
out-of-bounds memory access
```

The gradient tensors should be materialized with `.contiguous()` before the kernel launch, or their actual strides should be passed to and respected by the Triton kernels.
